### DISCUSSION TITLE
Technical Differences between SuperSimpleTCP and CavemanTCP

### DISCUSSION THREAD

**Discussion by dreamparkvr** [jchristn/CavemanTcp#11] [General]
Hey it's me again :)

I've been experimenting with CavemanTCP in some console apps and everything is going great. My use case is a fringe one so I didn't want to put this in Issues. 

I am trying to use CavemanTCP in a Unity3D application. I was able to successfully use SuperSimpleTCP and rig up a basic example without issue. But, doing the same now in CavemanTCP is giving me one issue I am having trouble figuring out. For some reason, when I close my application it locks up without an error. Everything else seems to work fine oddly enough.  

I'm not doing anything particularly crazy in my code. in fact its pretty much the same as what I was doing with SuperSimpleTCP aside from the missing event handler for data received. Strange because SuperSimple and Caveman seem to have pretty much the same DNA 🤷‍♂️ 

using CavemanTcp;
using System;
using UnityEngine;
using UnityEngine.UI;

public class Test : MonoBehaviour
{
    static CavemanTcpClient _Client;

    // Start is called before the first frame update
    void Start()
    {
        InitializeClient();
        Debug.Log("Connecting to tcp://127.0.0.1:8000");
        _Client.Connect(10);
    }

    static void InitializeClient()
    {
        // _Client = new CavemanTcpClient("127.0.0.1", 8000, false, null, null);
        _Client = new CavemanTcpClient("127.0.0.1:8000");

        _Client.Events.ClientConnected += (s, e) =>
        {
            Debug.Log("Connected to server");
            
        };

        _Client.Events.ClientDisconnected += (s, e) =>
        {
            Debug.Log("Disconnected from server");
        };
    }
}

**Comment by jchristn** [comment]
Hi @dreamparkvr when you say "when I close my application it locks up without an error" what exactly is happening?  Sorry I'm not super familiar with Unity3D.

**Reply by dreamparkvr** [reply]
No worries I understand! 
When I compile my sample app, it runs fine. Closing the application by rightclicking on it in the taskbar and closing it results in the application freezing and the only way to kill it entirely is from task manager. 


When I run my app just in the unity Editor, it runs perfectly fine until you stop the application, and then try to run it agian. It just hangs. 

Super strange. 

All I get out of the windows/application error log is The program CMTCPClientTest.exe version 2021.3.9.14448 stopped interacting with Windows and was closed, but nothing much else. 

Still trying to dig into other ways to debug it. I'm sure I'll figure something out. Just figured I'd leave a message hear in case you or anyone else has seen anything like this before. 

Like i said I just found it strange that SuperSimple works flawlessly 🤷

**Reply by jchristn** [reply]
Could you try setting the ```Logger``` property and catching the output to see if something pops up?  There is also ```Events.ExceptionEncountered``` which may be useful to see if something is happening.

**Reply by dreamparkvr** [reply]
I tried the logger but nothing aditional is coming out. 
When I close the app I can see that it's hitting Client.Disconnect() successfully. but after that, its almost as if there is an infinite loop or some underlying Thread(s) that aren't being terminated. 

The search is on :-P

**Reply by dreamparkvr** [reply]
I think I have it sorted (sort of). I reverted to 1.3.3.1 and for some reason everything is working fine. 

For now at least, I'll take it. Until i run into my next hurdle :-P

**Reply by dreamparkvr** [reply]
One more dumb question. 
Is there a good way to see if there are bytes to be read before calling read?

**Reply by jchristn** [reply]
Unfortunately network sockets (at least the implementation under the library) don't really provide this kind of capability. I wish....
